### ISSUE TITLE
Bug: Working with arrays/repeatedField from JSON to XML

### ISSUE THREAD

**Issue by TobiasNx** [metafacture/metafacture-fix#106]
If we move from JSON to XML. You can*t append to an array-field without `[]`

[Example](https://metafacture.org/playground/?data=%7B%0A++++%22animals%22%3A+%5B%0A++++++++%22dog%22%2C+%22elefant%22%0A++++%5D%0A%7D&flux=PG_DATA%0A%7C+as-records%0A%7C+decode-json%0A%7C+fix%0A%7C+encode-xml%0A%7C+print%0A%3B&fix=move_field%28%22animals%5B%5D%22%2C+%22animals%22%29%0Aadd_field%28%22dragon%22%2C+%22animals.%24append%22%29&active-editor=fix)

**Comment by blackwinter** [comment]
This seems to be working now, doesn't it? The `add_field()` syntax was wrong, must be `add_field("animals.$append", "dragon")`.

**Comment by TobiasNx** [comment]
You are right that the add_field had an error. But now it seems that the repeated field is created as subfield with index-numbers:
https://github.com/TobiasNx/fix-FunctionalReview-Testing/commit/ed3c4aeb0cce81119abd2edc5e7d25156d1a53c7


Should be: 

```
<?xml version="1.0" encoding="UTF-8"?>
<records>
	<record>
		<animals[]>dog</animals[]>
		<animals[]>elefant</animals[]>
		<animals>dog</animals>
		<animals>elefant</animals>
	</record>
</records>
```

is:
```
<?xml version="1.0" encoding="UTF-8"?>
<records>

	<record>
		<animals[]>
			<1>dog</1>
			<2>elefant</2>
		</animals[]>
		<animals>
			<1>dog</1>
			<2>elefant</2>
		</animals>
	</record>

</records>
```

**Comment by blackwinter** [comment]
> But now it seems that the repeated field is created as subfield with index-numbers

Yes, that's how arrays are emitted. It would be the responsibility of the encoder to produce a suitable output.

Are you arguing that arrays should always be emitted as repeated fields instead? Or should we take the `ARRAY_MARKER` into account?

**Comment by TobiasNx** [comment]
> Are you arguing that arrays should always be emitted as repeated fields instead?

Yes, I thought that fields should be emitted as they come into metafix. Especially if you only change small things in an record. For example xml -> xml and only add a field or filter out certain records. 


We should discuss this perhaps later. But I am not sure. Perhaps you are right that this should be something handled by the encoder.

**Comment by TobiasNx** [comment]
I think Fabian had something with the array-handler at the beginning. But I don't know.

**Comment by blackwinter** [comment]
> Yes, I thought that fields should be emitted as they come into metafix.

That's probably coloured by the fundamental difference between Morph and Fix. With Metamorph, you're producing a suitably named entity (that your designated encoder understands, e.g. `JsonEncoder`) with nested literals (whose names are then ignored by the `JsonEncoder`). With Metafix, you still have to name your fields appropriately but the entity/literal generation is done automatically. The difference now is that with Metamorph you're modeling the whole output, while with Metafix you're (usually) only manipulating ("fixing") your input in some way.

So, essentially, what you're proposing fits better with the Fix model, while it wouldn't integrate as easily with the Metafacture (Morph) model. But then again, that's probably `JsonEncoder`'s "fault", isn't it? (BTW: Are there any other encoders that have a concept of arrays?)

> We should discuss this perhaps later.

Not really. It's a pretty substantial question that should be answered rather sooner than later. Let me think about it a little more. And, of course, wait for @fsteeg's view on this matter.

**Comment by TobiasNx** [comment]
Your YAML does that too, i think

**Comment by blackwinter** [comment]
> Your YAML does that too, i think

Yeah, well, YAML and Json are basically the same ;) (YAML is a superset of Json.) But thanks for reminding me that we also have the `YamlEncoder` now :)

**Comment by blackwinter** [comment]
> Or should we take the `ARRAY_MARKER` into account?

I'm almost sure that's the way to go. If you're producing fields _marked_ as arrays, then they're emitted as indexed entities. Otherwise, arrays are just emitted as repeated fields. This way, the transformation (no matter if it's Morph or Fix) has full control over the output. WDYT?

**Comment by fsteeg** [comment]
> If you're producing fields marked as arrays, then they're emitted as indexed entities (for any downstream encoders that recognise this decoration and expect the corresponding structure). Otherwise, arrays are just emitted as repeated fields. This way, the transformation (no matter if it's Morph or Fix) has full control over the output. WDYT?

Sounds very good, and I thought this was actually (partially?) already happening. I/we had to change field names to end with `[]` to get correct output. But I guess that was only due to the way JsonEncoder handled the Metafix output.

**Comment by blackwinter** [comment]
So here's what it might look like: 8f69733. Only 79 more test failures to go :)

**Comment by fsteeg** [comment]
Looks good. Maybe we can avoid some work if we add `[]` for some cases instead of changing the expected output.

**Comment by blackwinter** [comment]
> Maybe we can avoid some work if we add `[]` for some cases instead of changing the expected output.

We could, but I assume those cases are intentionally bracket-less. The one I adapted here first wasn't really, because it doesn't matter for those tests. But most other cases probably meant to differentiate between our two kinds of arrays.

**Comment by fsteeg** [comment]
Functional review: @TobiasNx
Code review: @fsteeg

**Comment by TobiasNx** [comment]
https://github.com/TobiasNx/fix-FunctionalReview-Testing/blob/9a31d515b364259800b4458f3c4197996c5e6829/data/testing/repeatedFieldsArrayToXML

This seems to work. +1
But on the way i found that $first/$last is not working in this example. Will open another ticket for that.
